lib/model/model.go

```diff
@@ -2771,48 +2771,29 @@ func readOffsetIntoBuf(fs fs.Filesystem, file string, offset int64, buf []byte)
 // The exists function is expected to return true for all known paths
 // (excluding "" and ".")
 func unifySubs(dirs []string, exists func(dir string) bool) []string {
-	subs := trimUntilParentKnown(dirs, exists)
-	sort.Strings(subs)
-	return simplifySortedPaths(subs)
-}
-
-func trimUntilParentKnown(dirs []string, exists func(dir string) bool) []string {
-	var subs []string
-	for _, sub := range dirs {
-		for sub != "" && !fs.IsInternal(sub) {
-			sub = filepath.Clean(sub)
-			parent := filepath.Dir(sub)
-			if parent == "." || exists(parent) {
-				break
-			}
-			sub = parent
-			if sub == "." || sub == string(filepath.Separator) {
-				// Shortcut. We are going to scan the full folder, so we can
-				// just return an empty list of subs at this point.
-				return nil
-			}
-		}
-		if sub == "" {
-			return nil
-		}
-		subs = append(subs, sub)
+	if len(dirs) == 0 {
+		return nil
 	}
-	return subs
-}
-
-func simplifySortedPaths(subs []string) []string {
-	var cleaned []string
-next:
-	for _, sub := range subs {
-		for _, existing := range cleaned {
-			if sub == existing || strings.HasPrefix(sub, existing+string(fs.PathSeparator)) {
-				continue next
-			}
+	sort.Strings(dirs)
+	if dirs[0] == "" || dirs[0] == "." || dirs[0] == string(fs.PathSeparator) {
+		return nil
+	}
+	unified := dirs[:0]
+	prev := "./" // Anything that can't be parent of a clean path
+	for _, dir := range dirs {
+		dir = filepath.Clean(dir)
+		if dir == prev || strings.HasPrefix(dir, prev+string(fs.PathSeparator)) {
+			continue
 		}
-		cleaned = append(cleaned, sub)
+		parent := filepath.Dir(dir)
+		for parent != "." && parent != string(fs.PathSeparator) && !exists(parent) {
+			dir = parent
+			parent = filepath.Dir(dir)
+		}
+		unified = append(unified, dir)
```

Comment:
This is subtle and scary stuff as we're here modifying the backing array for dirs which we are also iterating over. It works because we are always looking at the same element we would be overwriting or one further along in the array - but it at the very least deserves a comment noting that this happens as it's liable to be a disaster if this gets refactored in some way. For example so that a given path gets split into two that both get appended or something.

Reply:
I see that now, I didn't even think of it as dangerous because using the same backing array was my initial thought for doing this. Instead of writing a comment, I rewrote the loop to explicitly modify the `dirs` slice given as argument. How does that look?